### PR TITLE
 ddl: add check for max index key length when alter modify/change column

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve? 
```sql
create table t1 (a varchar(10),b varchar(100),c tinyint,index(a),index(a,b),index (b,c));
alter table t1 modify column a varchar(3000); -- should return error.
```

MySQL
```sql
create table t1 (a varchar(10),b varchar(100),c tinyint,index(a),index(a,b),index (b,c));
alter table t1 modify column a varchar(3000);
(1071, u'Specified key was too long; max key length is 3072 bytes')
```

TiDB
```sql
create table t1 (a varchar(10),b varchar(100),c tinyint,index(a),index(a,b),index (b,c));
alter table t1 modify column a varchar(3000); -- execute sucessfully.

```

### What is changed and how it works?
* add check for max index key length when alter modify/change column

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by crazycs520** [comment]
/run-integration-common-test=pr/858

File: ddl/index.go [hunk 1]
```diff
@@ -137,6 +85,93 @@ func buildIndexColumns(columns []*model.ColumnInfo, idxColNames []*ast.IndexColN
...
+func getIndexColumnLength(col *model.ColumnInfo, colLen int) (int, error) {
+	// Take care of the sum of length of all index columns.
+	if colLen != types.UnspecifiedLength {
+		return colLen, nil
+	}
+	// Specified data types.
+	if col.Flen != types.UnspecifiedLength {
+		// Special case for the bit type.
+		if col.FieldType.Tp == mysql.TypeBit {
+			return (col.Flen + 7) >> 3, nil
```
Comment (AilinKid) [inline ddl/index.go]: I am not familiar with types，how does this `(col.Flen + 7) >> 3` expression come from?

Comment (crazycs520) [reply ddl/index.go]: The type is `bit`.

**Comment by tangenta** [comment]
What is the negative result of exceeding the max length(3072) in TiDB? Do we need to be compatible with MySQL's weakness?

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -2660,6 +2664,43 @@ func (d *ddl) getModifiableColumnJob(ctx sessionctx.Context, ident ast.Ident, or
 	return job, nil
 }
 
+// checkColumnWithIndexConstraint uses to check the related index constrain of the modified column.
+// index has a max prefix length constrain. eg: a varchar(100), index idx(a), then if modify column a to a varchar(4000),
+// that will cause index idx break the max prefix length constrain.
```
Comment (tangenta) [inline ddl/ddl_api.go]:
```suggestion
// checkColumnWithIndexConstraint is used to check the related index constraint of the modified column.
// index has a max-prefix-length constraint. eg: a varchar(100), index idx(a), modifying column a to a varchar(4000)
// will cause index idx to break the max-prefix-length constraint.
```

**Comment by imtbkcat** [comment]
@crazycs520 Please cherry-pick this pr to release-3.0.

**Pull request merged**